### PR TITLE
Issue #49

### PR DESCRIPTION
**Pull request by thongteav**
- Search and filter interface, user can search for a term, or filter with the options, media table now shows uploader and date as well.
image
- Searching returns any result of name or tag name matching the search term
image
- Filtering can apply on uploader, type, stage, tag, month and/or year
image
- Reset filters button only reset the filters and keep the search term
image

### CODE REVIEW

File: app/assets/javascripts/media.coffee [hunk 1]
```diff
@@ -1,3 +1,14 @@
...
+init = () ->
+	$('#reset_filters_btn').click ->
+    $('#uploader').get(0).selectedIndex = 0;
+    $('#type').get(0).selectedIndex = 0;
+    $('#stage').get(0).selectedIndex = 0;
+    $('#tag').get(0).selectedIndex = 0;
+    $('#month').get(0).selectedIndex = 0;
+    $('#year').get(0).selectedIndex = 0;
+    $('form').submit()
+
```
Comment (charlie0605) [inline app/assets/javascripts/media.coffee]: check above indentation

File: app/controllers/media_controller.rb [hunk 2]
```diff
@@ -1,7 +1,96 @@
...
+        med = avt + sd + bd
+      end
+      med.each do |m|
+          m.tags.each do |t|
+            if t.name.match(params[:term])
+              unless others.include? m
+                others << m
+              end
+            end
+          end
```
Comment (charlie0605) [inline app/controllers/media_controller.rb]: indentation here too

File: app/controllers/media_controller.rb [hunk 3]
```diff
@@ -1,7 +1,96 @@
...
+      end
+    end
+
+    if params[:type].present?
```
Comment (charlie0605) [inline app/controllers/media_controller.rb]: this if statement and the above if statement looks the same, do you need two or can you combine the two together?

File: app/views/media/_search.html.erb [hunk 4]
```diff
@@ -0,0 +1,53 @@
+<div>
+  <h4>Search and Filter:</h4>
+
+  <%= form_tag(media_path, method: :get) do %>
+    <div class="row">
+      <div class="field col-sm-2">
+        <%= label_tag :uploader %>
+        <%= select_tag :uploader, options_for_select(User.all.map { |user| [ user.nickname, user.id ] }.sort, :selected => params[:uploader]) , :class => "form-control", :prompt => 'Choose Uploader', :onchange => "this.form.submit();" %>
```
Comment (charlie0605) [inline app/views/media/_search.html.erb]: user.nickname has been changed to username, this will give an error

Comment (thongteav) [reply app/views/media/_search.html.erb]: You're right. Thanks for picking this up.

**Comment by thongteav** [comment]
@charlie0605 I've made the changes. Tested on chrome locally. Should work now. Please test it again.

File: app/controllers/media_controller.rb [hunk 5]
```diff
@@ -1,7 +1,68 @@
...
+
+    if params[:year].present?
+      med = Array.new
+      @media.each do |m|
+        if m.created_at.year.to_s.match(params[:year])
+          med << m
+        end
+      end
+      @media = med
+    end
```
Comment (thongteav) [inline app/controllers/media_controller.rb]: I tried to use database query with Model.where('extract(year  from date_column) = ?', desired_year), but SQLite doesn't support 'extract method' so I can't test it locally. That would be more tidy than looping it with ruby.
Not sure if there's any method that can extract the year for both SQL and SQLite.

**Comment by charlie0605** [comment]
the "reset filter" button should just be reset, and it should also reset the values inside the search input field.

The word Uploader sounds a bit weird, @helenvarley what's your preference? I thought maybe Owner?

Also having a horizontal line break just after the search and filter is good to separate the table.

The date is weird, its one day behind when I uploaded and the date format is in month, day and year.

overall should look something like this, what do you guys think, @rohrlach @helenvarley @vickis 
image

**Comment by thongteav** [comment]
@charlie0605 that's the same feature as in the current UpStage website. That's why instead of 'Reset' button, I intentionally put 'Reset Filters' on the button. But yeah, could change it if that's what we want.

**Comment by thongteav** [comment]
I could add a reset buttons that will clear the search term and reset the filters too if we want that too.

**Comment by charlie0605** [comment]
I think one button is enough 
@helenvarley @vickis @rohrlach whats your thoughts

**Comment by thongteav** [comment]
But I feel like 'Reset filters' is a bit more useful because to reset all, you could also just press the media button on the top.

**Comment by rohrlach** [comment]
We could just use 'Clear' like most apps do?

**Comment by rohrlach** [comment]
Clear *usually* clears everything.  I think if someone wanted to clear a specific filter, they'd just change that filter to 'All'

**Comment by charlie0605** [comment]
@helenvarley @vickis can we go ahead with changing the "uploader" to "Uploaded By", and leaving the reset filters as it is?

**Comment by thongteav** [comment]
@charlie0605 can you review this again? thanks.

**Comment by thongteav** [comment]
> The `created_at` and `updated_at` values are written from the server to the database, so they will always be in the server's or application's time zone. By the time the server processes the form data and saves to the database, it has no knowledge of the browser's time zone.

@alexanderglenn I guess it's for consistency, so that media items created at different time zone at one time would have the same created_at time. But I think we can configure the time zone difference.

@rohrlach do we need to configure the time zone so it creates the records at the uploader's time zone or just leave it as it is?

**Comment by thongteav** [comment]
@alexanderglenn It should show the right current time now. 
image

**Comment by helenvarley** [comment]
this actually merits a bit more discussion about the implications of using a player's local time or the server time. time is a big issue when you are working across multiple timezones! ;)

if it's the server timezone, then things can be easily kept in real chronological order. for example if vicki & i are both working on media for a show at the same time, me in europe & vi in nz, we will be able to see what are the most recent versions of our combined uploads. this could be useful. 
but if it's the player's local time, then i can see for example what i uploaded on a particular day, & have a record that is attached to my personal reality of time.

i am not sure which is better. both have pros & cons. until now it has always been the time of the server, & it has not been problematic.

definitely we want to be able to calculate in local time, for scheduling performances & rehearsals - this will be super useful if we can have a built-in time-converter. 

could players choose whether to use server time or local time for their uploads? or is this getting too complicated? (since it is really a minor issue & we don't want to lose time on it that could be devoted to more important things!)

**Comment by thongteav** [comment]
@helenvarley I've kept the time stored in the database uniform as it currently is. But the time displayed would be based on the time zone of the user. So say, person A and person B both uploaded file at the same time but in different time zones(maybe A is 5pm and B is 10pm), it should show player A both files uploaded at 5pm and show player B both files uploaded at 10pm.

**Comment by charlie0605** [comment]
@helenvarley I believe the time is uploaded as UTC. With the new code I think it will automatically determine your timezone and convert the UTC time to your local time. 

That means if I uploaded now (8:35PM NZT) it will reflect that on my end, and if you were to look at it at your end it would say 10:35AM (Munich Time). Is that something that you want?

**Comment by vickis** [comment]
The media upload at server time seems the most useful

a built in time converter would be fabulous - something like at the bottom
of the stage list page ....

**Comment by rohrlach** [comment]
Take a look at ```app/views/announcements/_announcement.html.erb``` line 5:
```
    <small><%= "#{localize(announcement.updated_at, :format => :long)} by #{announcement.author.username}" %></small>
```

**Comment by thongteav** [comment]
@rohrlach I've done `<td><%= m.created_at.localtime.strftime("%d/%m/%y @ %I:%M %p") %></td>` Is that ok?

File: app/controllers/media_controller.rb [hunk 6]
```diff
@@ -1,7 +1,72 @@
...
+    if params[:year].present?
+      if Rails.env.production?
+        avatars = Avatar.where('extract(year from avatars.created_at) = ?', params[:year])
+        backdrops = Backdrop.where('extract(year from backdrops.created_at) = ?', params[:year])
+        sounds = Sound.where('extract(year from sounds.created_at) = ?', params[:year])
+      else
+        avatars = Avatar.where("cast(strftime('%Y', avatars.created_at) as int) = ?", params[:year])
+        backdrops = Backdrop.where("cast(strftime('%Y', backdrops.created_at) as int) = ?", params[:year])
+        sounds = Sound.where("cast(strftime('%Y', sounds.created_at) as int) = ?", params[:year])
+      end
```
Comment (rohrlach) [inline app/controllers/media_controller.rb]: This seems antithetical to using an ORM.  ActiveRecord has platform independent querying ability.  Take another look at the documentation, we want as little raw SQL as possible.  In general, we don't want to be checking for environment type in the business logic, the application shouldn't be aware of whether it is in production or not.

File: app/controllers/media_controller.rb [hunk 7]
```diff
@@ -1,7 +1,72 @@
...
+        @media = backdrops
+      when "Sound"
+        @media = sounds
+      end
+    else
+      @media = avatars + sounds + backdrops
+    end
+
     @media.sort_by { |m| m.name }
   end
```
Comment (rohrlach) [inline app/controllers/media_controller.rb]: This method is now far too long - Consider ways in which you can split it apart, there could be opportunities to reuse code.

File: app/controllers/media_controller.rb (hunk 6 above)
Comment (thongteav) [reply app/controllers/media_controller.rb]: Sorry @rohrlach, I couldn't find a way to query the date with just the year or month. But I did find a gem that allows us to do that. [by_star gem](https://github.com/radar/by_star) Is it ok to use it? I got a warning when installing it though: 
image

Comment (rohrlach) [reply app/controllers/media_controller.rb]: Looks to be still in active development, so I don't see why not.

You should be able to address the issue by using the standard declaration in the Gemfile like so:
```
'by_star', '~> 3.0',
```

**Comment by rohrlach** [comment]
@thongteav What's the deal with this PR?

**Comment by thongteav** [comment]
@rohrlach I've made the changes as requested. So removing the checking of rails environment and querying with the gem installed. I've also separated the filtering functions from the media controller index. Should I get our team to test it out again if you're okay with the final code?

**Comment by thongteav** [comment]
@charlie0605 Ok, will sort this out today.

**Comment by thongteav** [comment]
I've fixed the bug and done a more thorough testing this time. @charlie0605 please check if there's still any bug. Thanks.

**Comment by charlie0605** [comment]
@rohrlach I've tested the filter and search and it now works fine.

**Pull request merged**